Goal: pick 4 USA men's Gymnasts. Gymnasts cannot compete on the same appartaus more than once. And each apparatus needs to be competed on 3 times each, then get the largest sum of the top 3 out of 4 scores on each apparats

explore all possible combinations of athletes and apparatuses, aiming to find the combination that maximizes the total score. and return the 4 athletes

# Tabu Search

### IAA

Selected Athletes:
['khoi young' 'asher hong' 'fred richard' 'yul moldauer']
 Score:
260.57900000000006

### Apparatus
Selected Athletes:
['frederick nathaniel richard' 'paul juda' 'khoi young' 'brody malone']
 Score:
253.19598690476192

---

# Randomized-Heuristics:

### IAA
Total Score: ['joshua karnes' 'ian lasic-ellis' 'asher hong' 'yul moldauer']
 Score: 254.94

### Apparatus
Total Score: ['khoi young' 'paul juda' 'brody malone' 'fuzzy benas']
 Score: 250.58868690476191

---


#  Variable Neighborhood Search (VNS):

## IAA

Selected Athletes: ['fuzzy benas', 'yul moldauer', 'asher hong', 'fred richard']
 Score: 257.34200000000004


## Apparatus
Selected Athletes: ['shane wiskus', 'khoi young', 'donnell whittenburg', 'brody malone']
 Score: 251.11566666666667

---


# brute-force search - Not working for using all apparatus 3times

### Apparatus
Selected Athletes:
['aidan cuy', 'asher hong', 'paul juda', 'yul moldauer']
 Score:
169.7672115079365

### IAA

Selected Athletes: ['vitaliy guimaraes', 'khoi young', 'asher hong', 'yul moldauer']  Score: 173.974

---



# VNS
## IAA - mean

Selected Athletes: ['michael jaroh', 'vitaliy guimaraes', 'yul moldauer', 'taylor burkhart']
Total Score: 249.964

# Randomized-Heuristics:


## IAA - mean
Best Selection: ['vitaliy guimaraes' 'yul moldauer' 'asher hong' 'taylor christopulos']
Best Score: 249.93350000000004


# Tabu Search

## IAA - mean
Selected Athletes:
['curran phillips' 'asher hong' 'vitaliy guimaraes' 'yul moldauer']
Total Score:
253.882


In [5]:
import pandas as pd
import numpy as np
from itertools import combinations
from pulp import LpProblem, LpVariable, lpSum, LpMaximize


In [6]:
df = pd.read_csv('../../Data/cleandata22-23/men22_23.csv')
# df1 = pd.read_csv('../../Cleaning/USOPC/usa_men_apparatus.csv')
df2=pd.read_csv('update_usa_apparatus.csv')
iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 

In [ ]:
usa = iaa24[iaa24['Nation'] == 'USA']


In [ ]:
usa.sort_values(by='AA', ascending=False)

In [ ]:
df=usa.copy()

In [ ]:

# # Updated data without brute-force search
# data = {
#     'Algorithm': ['Tabu Search', 'Tabu Search', 'Randomized-Heuristics', 'Randomized-Heuristics', 'Variable Neighborhood Search (VNS)', 'Variable Neighborhood Search (VNS)'],
#     'IAA/Apparatus': ['IAA', 'Apparatus', 'IAA', 'Apparatus', 'IAA', 'Apparatus'],
#     'Selected Athletes': [
#         ['khoi young', 'asher hong', 'fred richard', 'yul moldauer'],
#         ['frederick nathaniel richard', 'paul juda', 'khoi young', 'brody malone'],
#         ['joshua karnes', 'ian lasic-ellis', 'asher hong', 'yul moldauer'],
#         ['khoi young', 'paul juda', 'brody malone', 'fuzzy benas'],
#         ['fuzzy benas', 'yul moldauer', 'asher hong', 'fred richard'],
#         ['shane wiskus', 'khoi young', 'donnell whittenburg', 'brody malone']
#     ],
#     'Total Score': [260.57900000000006, 253.19598690476192, 254.94, 250.58868690476191, 257.34200000000004, 251.11566666666667]
# }

# # Create DataFrame
# result_df = pd.DataFrame(data)

# # Add columns for each athlete
# athletes = list(set(athlete for athletes_list in result_df['Selected Athletes'] for athlete in athletes_list))

# for athlete in athletes:
#     result_df[athlete] = result_df['Selected Athletes'].apply(lambda x: 1 if athlete in x else 0)

# # Calculate the sum for each athlete column
# athlete_sums = result_df.groupby('Algorithm')[athletes].sum()



#  Variable Neighborhood Search (VNS):

## IAA - max

Selected Athletes: ['fuzzy benas', 'yul moldauer', 'asher hong', 'fred richard']
Total Score: 257.34200000000004

## IAA - mean

Selected Athletes: ['michael jaroh', 'vitaliy guimaraes', 'yul moldauer', 'taylor burkhart']
Total Score: 249.964

---
## Apparatus - max
Selected Athletes: ['shane wiskus', 'khoi young', 'donnell whittenburg', 'brody malone']
Total Score: 251.11566666666667


---

## IAA

In [ ]:
# iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
# usa = iaa24[iaa24['Nation'] == 'USA']

# usa.fillna(0, inplace=True)

# # Group by 'Athlete' and get maximum value for each athlete
# df_max = usa.groupby('Athlete').max()

# # Reset index to make 'Athlete' a column again
# df_max.reset_index(inplace=True)
# df=df_max.copy()
# df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])

In [57]:
import pandas as pd

iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and calculate mean value for each athlete
df_mean = usa.groupby('Athlete').mean()

# Reset index to make 'Athlete' a column again
df_mean.reset_index(inplace=True)
df = df_mean.copy()
df = df.drop(columns=['Rank',  'year', 'AA'])
# 'round', 'Nation',
df.head()

/var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ipykernel_15505/3032695639.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  usa.fillna(0, inplace=True)
/var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ipykernel_15505/3032695639.py:9: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_mean = usa.groupby('Athlete').mean()


Athlete      FX      PH       SR       VT       PB     HB
0       alex diab  13.225   0.000  14.5715  13.2760   0.0000   0.00
1  anthony koppie  11.900   4.650  12.5500  13.9500  11.7500  11.40
2      asher hong  13.116  12.728  14.3495  15.7175  14.5275  12.05
3       blake sun   0.000  13.300   0.0000   0.0000  13.2220   0.00
4    brandon dang   0.000  14.466   0.0000   0.0000   0.0000   0.00

In [58]:
# VNS Algorithm with Iteration through random_state values
def variable_neighborhood_search(df):
    apparatus_order = ['FX', 'PH', 'SR', 'VT', 'PB', 'HB']  

    def calculate_total_score(athletes_df):
        
        # Sum top 3 scores for each apparatus
        total_score = athletes_df[['FX', 'PH', 'SR', 'VT', 'PB', 'HB']].apply(lambda x: x.nlargest(3).sum()).sum()
        return total_score

    best_athletes = None
    best_score = float('-inf')
    best_seed = None

    for seed in range(6000): 
        
        # Randomly select 4 athletes with current seed
        selected_athletes = df.sample(4, random_state=seed)

        # Check if each apparatus is competed on 3 times
        if all(selected_athletes[apparatus].count() >= 3 for apparatus in apparatus_order):
            
            # total score for current solution
            current_score = calculate_total_score(selected_athletes)

            # Compare with current best solution
            if current_score > best_score:
                best_athletes = selected_athletes['Athlete'].tolist()
                best_score = current_score
                best_seed = seed

    return best_athletes, best_score, best_seed

# Run the VNS algorithm with different random_state values
best_athletes, total_score, best_seed = variable_neighborhood_search(df)

print("Selected Athletes:", best_athletes)
print("Total Score:", total_score)
print("Best Seed:", best_seed)


Selected Athletes: ['michael jaroh', 'vitaliy guimaraes', 'yul moldauer', 'taylor burkhart']
Total Score: 249.964
Best Seed: 2304


## Apparatus
Selected Athletes:
['colt walker' 'paul juda' 'khoi young' 'frederick nathaniel richard']
Total Score:
170.307825



In [7]:
df = df2.copy()

In [8]:
df.sort_values(by='name', inplace=True)


In [ ]:
df.head()

In [ ]:
# VNS Algorithm with Iteration through random_state values
def variable_neighborhood_search(df):
    apparatus_order = ['fx', 'ph', 'sr', 'vt', 'pb', 'hb']  # Define the order of apparatus

    def calculate_total_score(athletes_df):
       
        # Sum top 3 scores for each apparatus
        total_score = athletes_df[['fx', 'ph', 'sr', 'vt', 'pb', 'hb']].apply(lambda x: x.nlargest(3).sum()).sum()
        return total_score

    best_athletes = None
    best_score = float('-inf')
    best_seed = None

    for seed in range(7000):  
        
        # Randomly select 4 athletes with current seed
        selected_athletes = df.sample(4, random_state=seed)

        # Check if each apparatus is competed on 3 times
        if all(selected_athletes[apparatus].count() >= 3 for apparatus in apparatus_order):
            
            # total score for  current solution
            current_score = calculate_total_score(selected_athletes)

            # Compare with current best solution
            if current_score > best_score:
                best_athletes = selected_athletes['name'].tolist()
                best_score = current_score
                best_seed = seed

    return best_athletes, best_score, best_seed

# Run VNS algorithm with different random_state values
best_athletes, total_score, best_seed = variable_neighborhood_search(df)

print("Selected Athletes:", best_athletes)
print("Total Score:", total_score)
print("Best Seed:", best_seed)


In [ ]:
# Selected Athletes: ['donnell whittenburg', 'ian lasic-ellis', 'paul juda', 'asher hong']
# Total Score: 249.27948650793653
# Best Seed: 749


# Selected Athletes: ['shane wiskus', 'khoi young', 'donnell whittenburg', 'brody malone']
# Total Score: 251.11566666666667
# Best Seed: 2981


# Randomized-Heuristics:

## IAA - max
Best Selection: ['joshua karnes' 'ian lasic-ellis' 'asher hong' 'yul moldauer']
Best Score: 254.94

## IAA - mean
Best Selection: ['vitaliy guimaraes' 'yul moldauer' 'asher hong' 'taylor christopulos']
Best Score: 249.93350000000004
Best seed: 6346


## Apparatus
Best Selection: ['khoi young' 'paul juda' 'brody malone' 'fuzzy benas']
Best Score: 250.58868690476191

---

## IAA

In [ ]:
# iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
# usa = iaa24[iaa24['Nation'] == 'USA']

# usa.fillna(0, inplace=True)

# # Group by 'Athlete' and get maximum value for each athlete
# df_max = usa.groupby('Athlete').max()

# # Reset index to make 'Athlete' a column again
# df_max.reset_index(inplace=True)
# df=df_max.copy()
# df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])

# mean

In [32]:
import pandas as pd

iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and calculate mean value for each athlete
df_mean = usa.groupby('Athlete').mean()

# Reset index to make 'Athlete' a column again
df_mean.reset_index(inplace=True)
df = df_mean.copy()
df = df.drop(columns=['Rank',  'year', 'AA'])
# 'round', 'Nation',
df.head()

/var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ipykernel_15505/3032695639.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  usa.fillna(0, inplace=True)
/var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ipykernel_15505/3032695639.py:9: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_mean = usa.groupby('Athlete').mean()


Athlete      FX      PH       SR       VT       PB     HB
0       alex diab  13.225   0.000  14.5715  13.2760   0.0000   0.00
1  anthony koppie  11.900   4.650  12.5500  13.9500  11.7500  11.40
2      asher hong  13.116  12.728  14.3495  15.7175  14.5275  12.05
3       blake sun   0.000  13.300   0.0000   0.0000  13.2220   0.00
4    brandon dang   0.000  14.466   0.0000   0.0000   0.0000   0.00

In [53]:
import pandas as pd
import numpy as np

# Set the number of athletes to select
num_athletes_to_select = 4

# Apparatuses
apparatuses = ['FX', 'PH', 'SR', 'VT', 'PB', 'HB']

# Variables to store the best solution
best_score = 0
best_selection = None
best_seed = 0

# Number of iterations for trying different random seeds
num_iterations = 1000

# Iterate through a range of seed values
for num_i in range(0, 8001):
    # Set a seed for reproducibility
    np.random.seed(num_i)

    # Randomly select 4 athletes
    selected_athletes = np.random.choice(df['Athlete'], size=num_athletes_to_select, replace=False)

    # Calculate the score for the current selection
    scores = df[df['Athlete'].isin(selected_athletes)][apparatuses].sum(axis=1).nlargest(3).sum()

    # Update the best solution if the current one is better
    if scores > best_score:
        best_score = scores
        best_selection = selected_athletes
        best_seed=num_i

print("Best Selection:", best_selection)
print("Best Score:", best_score)
print("Best seed:", best_seed)


Best Selection: ['vitaliy guimaraes' 'yul moldauer' 'asher hong' 'taylor christopulos']
Best Score: 249.93350000000004
Best seed: 6346


## Apparatus
Best Selection: ['khoi young' 'paul juda' 'brody malone' 'fuzzy benas']
Best Score: 250.58868690476191

In [ ]:
df = df2.copy()

In [ ]:
df.info()

In [ ]:
import pandas as pd
import numpy as np

# number of athletes to select
num_athletes_to_select = 4

# apparatuses
apparatuses = ['fx', 'ph', 'sr', 'vt', 'pb', 'hb']

# variables to store best solution
best_score = 0
best_selection = None

# number of iterations for the Randomized-heuristic
num_iterations = 1000

for _ in range(num_iterations):
   
    # Randomly select 4 athletes
    selected_athletes = np.random.choice(df['name'], size=num_athletes_to_select, replace=False)

    # score for the current selection
    scores = df[df['name'].isin(selected_athletes)][apparatuses].sum(axis=1).nlargest(3).sum()

    # Update best solution if the current one is better
    if scores > best_score:
        best_score = scores
        best_selection = selected_athletes

print("Best Selection:", best_selection)
print("Best Score:", best_score)


# Tabu Search

## IAA

Selected Athletes:
['khoi young' 'asher hong' 'fred richard' 'yul moldauer']
Total Score:
260.57900000000006

## IAA - mean

Selected Athletes:
['curran phillips' 'asher hong' 'vitaliy guimaraes' 'yul moldauer']
Total Score:
253.882

## Apparatus
Selected Athletes:
['frederick nathaniel richard' 'paul juda' 'khoi young' 'brody malone']
Total Score:
253.19598690476192

---

## IAA

In [31]:
# iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
# usa = iaa24[iaa24['Nation'] == 'USA']

# usa.fillna(0, inplace=True)

# # Group by 'Athlete' and get maximum value for each athlete
# df_max = usa.groupby('Athlete').max()

# # Reset the index to make 'Athlete' a column again
# df_max.reset_index(inplace=True)
# df=df_max.copy()
# df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])
# df.head()

# Mean

In [20]:
import pandas as pd

iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and calculate mean value for each athlete
df_mean = usa.groupby('Athlete').mean()

# Reset index to make 'Athlete' a column again
df_mean.reset_index(inplace=True)
df = df_mean.copy()
df = df.drop(columns=['Rank',  'year', 'AA'])
# 'round', 'Nation',
df.head()

/var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ipykernel_15505/3032695639.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  usa.fillna(0, inplace=True)
/var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ipykernel_15505/3032695639.py:9: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_mean = usa.groupby('Athlete').mean()


Athlete      FX      PH       SR       VT       PB     HB
0       alex diab  13.225   0.000  14.5715  13.2760   0.0000   0.00
1  anthony koppie  11.900   4.650  12.5500  13.9500  11.7500  11.40
2      asher hong  13.116  12.728  14.3495  15.7175  14.5275  12.05
3       blake sun   0.000  13.300   0.0000   0.0000  13.2220   0.00
4    brandon dang   0.000  14.466   0.0000   0.0000   0.0000   0.00

In [21]:
import pandas as pd
import numpy as np

def calculate_score(selected_athletes, df):
    """
    Calculate the total score for a given set of selected athletes.
    """
    # Select rows corresponding to the chosen athletes
    selected_df = df[df['Athlete'].isin(selected_athletes)]

    # um of top 3 scores for each apparatus
    apparatus_scores = selected_df.drop(['Athlete'], axis=1)
    top_3_sum = apparatus_scores.apply(lambda row: sum(sorted(row, reverse=True)[:3]), axis=0)

    # total score
    total_score = top_3_sum.sum()

    return total_score




def tabu_search(df, num_athletes_to_select, tabu_size, max_iterations):
    
    # Initialize solution with a random set of athletes
    available_athletes = df['Athlete'].tolist()
    current_solution = np.random.choice(available_athletes, size=num_athletes_to_select, replace=False)
    best_solution = current_solution.copy()
    best_score = calculate_score(best_solution, df)

    # Initialize tabu list
    tabu_list = []

    # Main Tabu Search loop
    for iteration in range(max_iterations):
        
        # Generate neighbors by swapping athletes
        neighbors = []
        for i in range(num_athletes_to_select):
            for athlete in df['Athlete']:
                if athlete not in current_solution:
                    neighbor = current_solution.copy()
                    neighbor[i] = athlete
                    neighbors.append(neighbor)
                    
                best_neighbor = None
        best_neighbor_score = float('-inf')

        for neighbor in neighbors:
            neighbor_score = calculate_score(neighbor, df)

            # Convert the NumPy array to a tuple
            neighbor_tuple = tuple(neighbor)

            if neighbor_score > best_neighbor_score and neighbor_tuple not in tabu_list:
                best_neighbor = neighbor
                best_neighbor_score = neighbor_score

        # Update current solution and tabu list
        current_solution = best_neighbor
        tabu_list.append(tuple(best_neighbor))

        # Keep tabu list size within specified limit
        if len(tabu_list) > tabu_size:
            tabu_list.pop(0)

        # Update best solution if a better one is found
        if best_neighbor_score > best_score:
            best_solution = best_neighbor
            best_score = best_neighbor_score

        # Print progress
        print(f"Iteration {iteration + 1}, Best Score: {best_score}")

    return best_solution, best_score
    
# Set parameters
num_athletes_to_select = 4
tabu_size = 10
max_iterations = 100

# Perform Tabu Search
best_solution, best_score = tabu_search(df, num_athletes_to_select, tabu_size, max_iterations)

print('Selected Athletes:')
print(best_solution)
print('Total Score:')
print(best_score)

      

Iteration 1, Best Score: 246.57199999999997
Iteration 2, Best Score: 250.36700000000002
Iteration 3, Best Score: 252.44350000000003
Iteration 4, Best Score: 253.882
Iteration 5, Best Score: 253.882
Iteration 6, Best Score: 253.882
Iteration 7, Best Score: 253.882
Iteration 8, Best Score: 253.882
Iteration 9, Best Score: 253.882
Iteration 10, Best Score: 253.882
Iteration 11, Best Score: 253.882
Iteration 12, Best Score: 253.882
Iteration 13, Best Score: 253.882
Iteration 14, Best Score: 253.882
Iteration 15, Best Score: 253.882
Iteration 16, Best Score: 253.882
Iteration 17, Best Score: 253.882
Iteration 18, Best Score: 253.882
Iteration 19, Best Score: 253.882
Iteration 20, Best Score: 253.882
Iteration 21, Best Score: 253.882
Iteration 22, Best Score: 253.882
Iteration 23, Best Score: 253.882
Iteration 24, Best Score: 253.882
Iteration 25, Best Score: 253.882
Iteration 26, Best Score: 253.882
Iteration 27, Best Score: 253.882
Iteration 28, Best Score: 253.882
Iteration 29, Best Score

## Apparatus
Selected Athletes:
['frederick nathaniel richard' 'paul juda' 'khoi young' 'brody malone']
Total Score:
253.19598690476192

In [ ]:
df=df2.copy()

In [ ]:
import pandas as pd
import numpy as np

def calculate_score(selected_athletes, df):
    """
    Calculate the total score for a given set of selected athletes.
    """
    # Select rows corresponding to chosen athletes
    selected_df = df[df['name'].isin(selected_athletes)]

    # sum of top 3 scores for each apparatus
    apparatus_scores = selected_df.drop([ 'name'], axis=1)
    top_3_sum = apparatus_scores.apply(lambda row: sum(sorted(row, reverse=True)[:3]), axis=0)

    # total score
    total_score = top_3_sum.sum()

    return total_score

def tabu_search(df, num_athletes_to_select, tabu_size, max_iterations):
    
    # Initialize solution with a random set of athletes
    available_athletes = df['name'].tolist()
    current_solution = np.random.choice(available_athletes, size=num_athletes_to_select, replace=False)
    best_solution = current_solution.copy()
    best_score = calculate_score(best_solution, df)

    # Initialize tabu list
    tabu_list = []

    # Main Tabu Search loop
    for iteration in range(max_iterations):
        
        # Generate neighbors by swapping athletes
        neighbors = []
        for i in range(num_athletes_to_select):
            for athlete in df['name']:
                if athlete not in current_solution:
                    neighbor = current_solution.copy()
                    neighbor[i] = athlete
                    neighbors.append(neighbor)               
                    
                best_neighbor = None
        best_neighbor_score = float('-inf')

        for neighbor in neighbors:
            neighbor_score = calculate_score(neighbor, df)

            # Convert the NumPy array to a tuple
            neighbor_tuple = tuple(neighbor)

            if neighbor_score > best_neighbor_score and neighbor_tuple not in tabu_list:
                best_neighbor = neighbor
                best_neighbor_score = neighbor_score

        # Update current solution and tabu list
        current_solution = best_neighbor
        tabu_list.append(tuple(best_neighbor))

        # Keep tabu list size within specified limit
        if len(tabu_list) > tabu_size:
            tabu_list.pop(0)

        # Update the best solution if a better one is found
        if best_neighbor_score > best_score:
            best_solution = best_neighbor
            best_score = best_neighbor_score

        # Print progress
        print(f"Iteration {iteration + 1}, Best Score: {best_score}")

    return best_solution, best_score

# Set parameters
num_athletes_to_select = 4
tabu_size = 300
max_iterations = 100

# Perform Tabu Search
best_solution, best_score = tabu_search(df, num_athletes_to_select, tabu_size, max_iterations)

print('Selected Athletes:')
print(best_solution)
print('Total Score:')
print(best_score)

      

## IAA

# brute-force search

## Apparatus
Selected Athletes:
['aidan cuy', 'asher hong', 'paul juda', 'yul moldauer']
Maximum Score:
169.7672115079365

## IAA

Selected Athletes: ['vitaliy guimaraes', 'khoi young', 'asher hong', 'yul moldauer'] Maximum Score: 173.974


---

## Apparatus


In [ ]:
df=df2.copy()

In [ ]:
df.head()

Over 3 hours to run
100%|█████████████████████████████████| 424270/424270 [3:13:04<00:00, 36.62it/s]

Selected Athletes:
['aidan cuy', 'asher hong', 'paul juda', 'yul moldauer']
Maximum Score:
169.7672115079365


In [ ]:
df.info()

## IAA

Selected Athletes:
['vitaliy guimaraes', 'khoi young', 'asher hong', 'yul moldauer']
Maximum Score:
173.974

In [ ]:
import pandas as pd
import itertools
from tqdm import tqdm

# Generate all possible combinations of athletes and apparatuses
athletes = df['name'].unique()
apparatuses = df.columns[1:]  # Exclude the 'name' column
combinations = list(itertools.product(athletes, apparatuses))

# variables to store best combination and maximum total score
best_combination = None
max_total_score = 0

# Iterate through all combinations with tqdm for progress bar
for combo in tqdm(itertools.permutations(combinations, 4), total=len(list(itertools.permutations(combinations, 4)))):
    
    # Create a DataFrame with selected combination
    selected_df = pd.DataFrame(list(combo), columns=['name', 'Apparatus'])
    
    # each athlete competes on each apparatus at most once
    if selected_df.duplicated(subset='name').any() or selected_df.duplicated(subset='Apparatus').any():
        continue
    
    # each apparatus is competed on 3 times
    if selected_df['Apparatus'].value_counts().min() < 3:
        continue
    
    # total score for current combination
    total_score = 0
    for _, row in selected_df.iterrows():
        total_score += df.loc[df['name'] == row['name'], row['Apparatus']].max()
    
    # Update best combination if total score is higher
    if total_score > max_total_score:
        max_total_score = total_score
        best_combination = selected_df

print("\nBest Combination:")
print(best_combination)
print("Maximum Total Score:", max_total_score)


In [ ]:
iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and get maximum value for each athlete
df_max = usa.groupby('Athlete').max()

# Reset index to make 'Athlete' a column again
df_max.reset_index(inplace=True)
df=df_max.copy()
df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])

df.head()

---

# Simulated Annealing
- Simulated annealing involves randomness in accepting worse solutions to explore the search space effectively. After adding Seed the results are terrible


Selected Athletes: ['fuzzy benas', 'yul moldauer', 'asher hong', 'fred richard']
Maximum Score: 174.61700000000002



In [ ]:
iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and get maximum value for each athlete
df_max = usa.groupby('Athlete').max()

# Reset index to make 'Athlete' a column again
df_max.reset_index(inplace=True)
df=df_max.copy()
df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])

df.head()

In [ ]:
import pandas as pd
import numpy as np

def calculate_score(selected_athletes, df):
    total_score = df[df['Athlete'].isin(selected_athletes)].iloc[:, 2:8].values
    total_score.sort(axis=1, kind='quicksort')
    return total_score[:, -3:].sum()

def simulated_annealing(df, num_athletes=4, iterations=1000, initial_temperature=1.0, cooling_rate=0.95):
    np.random.seed(88)  

    current_solution = np.random.choice(df['Athlete'], size=num_athletes, replace=False).tolist()
    current_score = calculate_score(current_solution, df)

    best_solution = current_solution.copy()
    best_score = current_score

    temperature = initial_temperature

    for iteration in range(iterations):
        neighbor_solution = current_solution.copy()
        idx1, idx2 = np.random.choice(num_athletes, size=2, replace=False)
        neighbor_solution[idx1], neighbor_solution[idx2] = neighbor_solution[idx2], neighbor_solution[idx1]

        neighbor_score = calculate_score(neighbor_solution, df)

        if neighbor_score > current_score or np.random.rand() < np.exp((neighbor_score - current_score) / temperature):
            current_solution = neighbor_solution
            current_score = neighbor_score

        if current_score > best_score:
            best_solution = current_solution.copy()
            best_score = current_score

        temperature *= cooling_rate

    return best_solution, best_score


# Run simulated annealing
best_athletes, best_score = simulated_annealing(df)

print('Selected Athletes:', best_athletes)
print('Maximum Score:', best_score)


# Harmony Search algorithm = results for 6 and not 4 athletes

              Athlete     FX      PH     SR     VT      PB     HB
715    patrick hoopes   0.00  14.992   0.00   0.00   0.000   0.00
708   curran phillips   0.00   0.000   0.00   0.00  15.997  13.70
427   ian lasic-ellis  13.65  13.500  13.75  13.55  13.300  13.30
711   spencer goodell  13.45   0.000   0.00   0.00  13.750   0.00
442      noah newfeld  12.20  13.050  12.10  13.50  13.200  12.15
444  garrett braunton  12.85  12.900  13.05  12.20  13.150  11.65

In [ ]:
iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and get maximum value for each athlete
df_max = usa.groupby('Athlete').max()

# Reset the index to make 'Athlete' a column again
df_max.reset_index(inplace=True)
df=df_max.copy()
df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])

df.head()

In [ ]:


import pandas as pd
import numpy as np

def fitness(solution, df):
    selected_athletes = df.iloc[solution]
    
    # Check if each apparatus is competed exactly three times
    apparatus_counts = selected_athletes.groupby('Athlete').size()
    if not all(count == 3 for count in apparatus_counts):
        return np.inf
    
    # Check if the selected athletes are unique for each apparatus
    if len(set(solution)) < len(solution):
        return np.inf
    
    scores = selected_athletes[['FX', 'PH', 'SR', 'VT', 'PB', 'HB']].values
    top3_sum = np.sum(np.partition(scores, -3)[:, -3:], axis=1)

    # egative because Harmony Search seeks to minimize objective function
    return -np.sum(top3_sum)  

def harmony_search(df, num_iterations, harmony_memory_size):
    num_athletes = len(df)
    
    # 6 apparatus (FX, PH, SR, VT, PB, HB)
    num_apparatus = 6  
    hm_size = harmony_memory_size

    # Initialize harmony memory with random solutions
    harmony_memory = np.random.randint(0, num_athletes, size=(hm_size, num_apparatus))

    for iteration in range(num_iterations):
        
        # Create new harmony solutions
        new_solution = np.random.randint(0, num_athletes, size=num_apparatus)

        # Update harmony memory with new solution
        worst_index = np.argmin([fitness(harmony, df) for harmony in harmony_memory])
        
        # Check if new solution is valid before updating memory
        if fitness(new_solution, df) < np.inf:
            harmony_memory[worst_index] = new_solution

    #best solution in harmony memory
    best_solution = harmony_memory[np.argmin([fitness(harmony, df) for harmony in harmony_memory])]

    return df.iloc[best_solution]

# Set number of iterations and harmony memory size
num_iterations = 100
harmony_memory_size = 10

# Run Harmony Search algorithm
selected_athletes = harmony_search(df, num_iterations, harmony_memory_size)

# Print selected athletes
print(selected_athletes[['Athlete', 'FX', 'PH', 'SR', 'VT', 'PB', 'HB']])


# Stochastic control

## IAA

In [ ]:
iaa24 = pd.read_csv('../../Combine_Data/men/iaa2024.csv') 
usa = iaa24[iaa24['Nation'] == 'USA']

usa.fillna(0, inplace=True)

# Group by 'Athlete' and get maximum value for each athlete
df_max = usa.groupby('Athlete').max()

# Reset index to make 'Athlete' a column again
df_max.reset_index(inplace=True)
df=df_max.copy()
df=df.drop(columns=['Rank', 'AA', 'round', 'Nation', 'year'])

In [ ]:
df.columns

In [ ]:
import pandas as pd
from itertools import combinations
from tqdm import tqdm

# Get all possible combinations of 4 gymnasts out of 5
gymnast_combinations = list(combinations(df['Athlete'], 4))

# variables to keep track of best result
best_result = None
best_sum = float('-inf')

# Iterate over all combinations of gymnasts with tqdm progress bar
for combination in tqdm(gymnast_combinations, desc="Combinations"):
    
    # rows corresponding to current combination of gymnasts
    selected_gymnasts = df[df['Athlete'].isin(combination)]
    
    # each apparatus is competed on 3 times by each selected gymnast
    selected_gymnasts = selected_gymnasts.groupby('Athlete').apply(lambda x: x.sample(n=3, replace=True, random_state=42)).reset_index(drop=True)
    
    # sum of top 3 scores for each apparatus
    selected_gymnasts['Top3Sum'] = selected_gymnasts[['FX', 'PH', 'SR', 'VT', 'PB', 'HB']].apply(lambda x: sum(sorted(x, reverse=True)[:3]), axis=1)
    
    # largest sum for each apparatus
    current_sum = selected_gymnasts.groupby('Athlete').agg({'Top3Sum': 'max'}).reset_index()['Top3Sum'].sum()
    
    # Update best result if the current combination is better
    if current_sum > best_sum:
        best_sum = current_sum
        best_result = selected_gymnasts[['Athlete', 'Top3Sum']].copy()

print(best_result)


In [ ]:
41.80+71.95+77.55+76.25

## Apparatus

In [ ]:
df=df2.copy()

In [ ]:
df

In [ ]:
import pandas as pd
from itertools import combinations
from tqdm import tqdm

# Get all possible combinations of 4 gymnasts out of 5
gymnast_combinations = list(combinations(df['name'], 4))

# variables to keep track of the best result
best_result = None
best_sum = float('-inf')

# Iterate over all combinations of gymnasts with tqdm progress bar
for combination in tqdm(gymnast_combinations, desc="Combinations"):
   
    # Select rows corresponding to the current combination of gymnasts
    selected_gymnasts = df[df['name'].isin(combination)]
    
    # each apparatus is competed on 3 times by each selected gymnast
    selected_gymnasts = selected_gymnasts.groupby('name').apply(lambda x: x.sample(n=3, replace=True, random_state=42)).reset_index(drop=True)
    
    # the sum of top 3 scores for each apparatus
    selected_gymnasts['Top3Sum'] = selected_gymnasts[['fx', 'hb', 'pb', 'ph', 'sr', 'vt']].apply(lambda x: sum(sorted(x, reverse=True)[:3]), axis=1)
    
    # largest sum for each apparatus
    current_sum = selected_gymnasts.groupby('name').agg({'Top3Sum': 'max'}).reset_index()['Top3Sum'].sum()
    
    # Update best result if the current combination is better
    if current_sum > best_sum:
        best_sum = current_sum
        best_result = selected_gymnasts[['name', 'Top3Sum']].copy()

print(best_result)


In [ ]:
78.149667+81.309627+84.374795+39.050000

In [ ]:
6*14.00